contracts/Marketplace.sol

```diff
@@ -0,0 +1,76 @@
+pragma solidity ^0.4.23;
+
+import './abstractions/MarketplaceTemplate.sol';
+import './common/Destructible.sol';
+import './libs/ProductLib.sol';
+
+contract Marketplace is MarketplaceTemplate, Destructible {
+    using ProductLib for ProductLib.Product;
+
+    bytes32[] public products;    
+
+    event LogNewProduct(bytes32 indexed id, string name, uint price, uint quantity);
+    event LogNewPurchase(bytes32 indexed id, uint quantity);
+    event LogStockUpdate(bytes32 indexed id, uint quantity);
+    event LogWithdrawal();
+
+    mapping (bytes32=>ProductLib.Product) public productById;
+
+    modifier productExists(bytes32 _id) {
+        require(productById[_id].exists());
+        _;
+    }
+
+    function buy(bytes32 _id, uint _quantity) public productExists(_id) payable {
+        require(productById[_id].getProductPrice(_quantity) <= msg.value);
+
+        emit LogNewPurchase(_id, _quantity);
+
+        productById[_id].purchaseProduct(_quantity);
+    }
+    
+    function update(bytes32 _id, uint _newQuantity) public productExists(_id) onlyOwner {
+        emit LogStockUpdate(_id, _newQuantity);
+
+        productById[_id].updateProduct(_newQuantity);
+    }
+    
+    // creates a new product and returns its ID
+    function newProduct(string _name, uint _price, uint _quantity) public onlyOwner returns(bytes32) {
+        require(_quantity > 0);
+        require(_price > 0);
+        require(bytes(_name).length > 0);
+
+        bytes32 id = keccak256(abi.encodePacked(_name, _price, _quantity, now));
+
+        require(!productById[id].exists());
+
+        products.push(id);
+        productById[id] = ProductLib.Product({name: _name, price: _price, quantity: _quantity});
+
+        emit LogNewProduct(id, _name, _price, _quantity);
+
+        return id;
```

Comment:
What is the purpose of that id? Can we get it when we invoke/sendTransaction? Is not that the purpose of the event? Check [this](https://ethereum.stackexchange.com/questions/3285/how-to-get-return-values-when-function-with-argument-is-called)

Reply:
I have no idea why that is there and I asked myself the same thing when I saw it. In the softuni project details it says that it is recommended that we should keep the provided functions the same, without touching their names, parameters and return values. Otherwise I agree it should not return anything.

Reply:
Let's keep it that way in order to satisfy the Softuni's interface.